src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/TextMapPropagator.cs

```diff
@@ -0,0 +1,520 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Text;
+using System.Net;
+using System.Collections.Generic;
+
+namespace System.Diagnostics
+{
+    public abstract class TextMapPropagator
+    {
+        public delegate bool PropagatorGetterCallback(object carrier, string fieldName, out string? value);
+
+        public abstract IReadOnlyCollection<string> Fields { get; }
+
+        // Inject
+
+        public abstract bool Inject(Activity activity, object carrier, Action<object /* carrier */, string /* field name */, string /* value to inject */> setter);
+        public abstract bool Inject(ActivityContext context, object carrier, Action<object /* carrier */, string /* field name */, string /* value to inject */> setter);
+        public abstract bool Inject(IEnumerable<KeyValuePair<string, string?>> baggage, object carrier, Action<object /* carrier */, string /* field name */, string /* value to inject */> setter);
+
+        // Extract
+
+        public abstract bool Extract(object carrier, PropagatorGetterCallback getter, out string? id, out string? state);
+        public abstract bool Extract(object carrier, PropagatorGetterCallback getter, out ActivityContext context);
+        public abstract bool Extract(object carrier, PropagatorGetterCallback getter, out IEnumerable<KeyValuePair<string, string?>>? baggage);
+
```

Comment:
I'm trying to avoid digging into the details of the API as long as it appears to satisfy the goals we wanted to achieve. I think this API can achieve it, but I have more faith in the practical demonstration than in my mental model : )

Reply:
We already trying the prototype with http client and asp.net to ensure no gaps or issues with that design.